File: mobile/app/src/main/java/ua/nure/cleaningservice/ui/contracts/SignContractActivity.java
package ua.nure.cleaningservice.ui.contracts;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.Spinner;

import androidx.appcompat.app.AppCompatActivity;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import ua.nure.cleaningservice.R;
import ua.nure.cleaningservice.data.Company;
import ua.nure.cleaningservice.data.Contract;
import ua.nure.cleaningservice.data.Placement;
import ua.nure.cleaningservice.data.Service;
import ua.nure.cleaningservice.network.JSONPlaceHolderApi;
import ua.nure.cleaningservice.network.NetworkService;

public class SignContractActivity extends AppCompatActivity {

    private static final String TAG = "SignContractActivity";

    private List<Contract> mContracts;
    private JSONPlaceHolderApi mApi;
    ImageView mBack;
    Button mSignContractButton;
    Spinner mServicesSpinner, mRoomsSpinner;
    Contract mContract;

    int roomId;
    int serviceId;
    int[] servicesId;
    List<Service> mServices;
    List<Placement> mPlacements;
    String token;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.sign_contract_activity);

        mBack = findViewById(R.id.back_btn);
        mSignContractButton = findViewById(R.id.sign_contract_btn);
        mSignContractButton.setOnClickListener(v -> signContract());

        mApi = NetworkService.getInstance().getApiService();
        mContract = new Contract();

        mServicesSpinner = findViewById(R.id.get_service_spinner);
        mServicesSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                serviceId = position;
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        mRoomsSpinner = findViewById(R.id.get_room_spinner);
        mRoomsSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                roomId = mPlacements.get(position).getId();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        getData();
    }

    Callback<ArrayList<Service>> serviceCallback = new Callback<ArrayList<Service>>() {
        @Override
        public void onResponse(Call<ArrayList<Service>> call, Response<ArrayList<Service>> response) {
            if (!response.isSuccessful()) {
                System.out.println(response.code());
                return;
            }
            mServices = response.body();
            String[] names = new String[mServices.size()];
            servicesId = new int[mServices.size()];
            int i = 0;
            for (Service service : mServices) {
                names[i] = service.getName();
                servicesId[i] = service.getId();
                i++;
            }
            initServicesAdapter(names);
        }

        @Override
        public void onFailure(Call<ArrayList<Service>> call, Throwable t) {
            System.out.println(t);
            Log.i(TAG, t.getMessage());
        }
    };

    Callback<ArrayList<Placement>> roomsCallback = new Callback<ArrayList<Placement>>() {
        @Override
        public void onResponse(Call<ArrayList<Placement>> call, Response<ArrayList<Placement>> response) {
            if (!response.isSuccessful()) {
                System.out.println(response.code());
                return;
            }
            mPlacements = response.body();
            Integer[] ids = new Integer[mPlacements.size()];
            int i = 0;
            for (Placement placement : mPlacements) {
                ids[i] = placement.getId();
                i++;
            }
            initRoomsAdapter(ids);
        }

        @Override
        public void onFailure(Call<ArrayList<Placement>> call, Throwable t) {
            System.out.println(t);
            Log.i(TAG, t.getMessage());
        }
    };

    Callback<Contract> signContractCallback = new Callback<Contract>() {
        @Override
        public void onResponse(Call<Contract> call, Response<Contract> response) {
            if (!response.isSuccessful()) {
                System.out.println(response.code());
                return;
            }
            finish();
        }

        @Override
        public void onFailure(Call<Contract> call, Throwable t) {
            System.out.println(t);
            Log.i(TAG, t.getMessage());
        }
    };

    private void getData() {
        String email = Company.getInstance().getEmail();
        token = "Bearer " + Company.getInstance().getToken();

        mApi.getServiceData(token, getIntent().getStringExtra("cEmail")).enqueue(serviceCallback);
        mApi.getPlacementData(token, email).enqueue(roomsCallback);
    }

    private void initServicesAdapter(String[] servicesData){
        ArrayAdapter<String> adapter = new ArrayAdapter(this,
                android.R.layout.simple_list_item_1, servicesData);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        mServicesSpinner.setAdapter(adapter);
    }

    private void initRoomsAdapter(Integer[] roomsData){
        ArrayAdapter<String> adapter = new ArrayAdapter(this,
                android.R.layout.simple_list_item_1, roomsData);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        mRoomsSpinner.setAdapter(adapter);
    }

    private void signContract() {
        mContract.setProviderServiceId(servicesId[serviceId]);
        mContract.setPlacementId(roomId);
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSX");
        mContract.setDate(dateFormat.format(new Date()));

            NetworkService.getInstance()
                    .getApiService()
                    .signContract(token, mContract)
                    .enqueue(signContractCallback);

    }

}

File: mobile/app/src/main/java/ua/nure/cleaningservice/ui/contracts/SignContractActivity.kt
package ua.nure.cleaningservice.ui.contracts

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.app.TimePickerDialog
import android.app.TimePickerDialog.OnTimeSetListener
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.format.DateUtils
import android.util.Log
import android.view.View
import android.widget.*
import android.widget.AdapterView.OnItemSelectedListener
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import ua.nure.cleaningservice.R
import ua.nure.cleaningservice.data.Contract
import ua.nure.cleaningservice.data.Placement
import ua.nure.cleaningservice.data.Service
import ua.nure.cleaningservice.data.User
import ua.nure.cleaningservice.network.JsonPlaceHolderApi
import ua.nure.cleaningservice.network.NetworkService
import ua.nure.cleaningservice.ui.util.LoadingDialog
import java.text.SimpleDateFormat
import java.util.*


class SignContractActivity : AppCompatActivity() {

    private var mApi: JsonPlaceHolderApi? = null
    lateinit var mBack: ImageView
    lateinit var mSignContractButton: Button
    lateinit var mServicesSpinner: Spinner
    lateinit var mRoomsSpinner: Spinner
    lateinit var timeButton: Button
    lateinit var dateButton: Button
    var mContract: Contract? = null
    var roomId = 0
    var serviceId = 0
    lateinit var servicesId: IntArray
    var mServices: List<Service>? = null
    var mPlacements: List<Placement>? = null
    var token: String? = null
    private val loadingDialog = LoadingDialog(this@SignContractActivity)

    @RequiresApi(Build.VERSION_CODES.N)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.sign_contract_activity)
        mBack = findViewById(R.id.back_btn)
        mBack.setOnClickListener{
            navigateToMenuScreen()
            finish()
        }

        mSignContractButton = findViewById(R.id.sign_contract_btn)
        mSignContractButton.setOnClickListener(View.OnClickListener { v: View? -> signContract() })
        mApi = NetworkService.getInstance().apiService
        mContract = Contract()
        mServicesSpinner = findViewById(R.id.get_service_spinner)
        mServicesSpinner.setOnItemSelectedListener(object : OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
                serviceId = position
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        })
        mRoomsSpinner = findViewById(R.id.get_room_spinner)
        mRoomsSpinner.setOnItemSelectedListener(object : OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
                roomId = mPlacements!![position].id
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        })

        timeButton = findViewById(R.id.timeButton)
        timeButton.setOnClickListener { v -> setTime(v) }

        dateButton = findViewById(R.id.dateButton)
        dateButton.setOnClickListener { v -> setDate(v) }

        currentDateTime= findViewById(R.id.currentDateTime)
        setInitialDateTime()
    }

    override fun onResume() {
        super.onResume()
        data
    }

    private val data: Unit
        get() {
            loadingDialog.start()
            val email = User.getInstance().email
            token = "Bearer " + User.getInstance().token
            mApi!!.getServiceData(token, intent.getStringExtra("cEmail")).enqueue(serviceCallback)
            mApi!!.getPlacementData(token, email).enqueue(roomsCallback)
        }

    var serviceCallback: Callback<ArrayList<Service>?> = object : Callback<ArrayList<Service>?> {
        override fun onResponse(call: Call<ArrayList<Service>?>, response: Response<ArrayList<Service>?>) {
            if (!response.isSuccessful) {
                println(response.code())
                loadingDialog.dismiss()
                return
            }
            mServices = response.body()
            val names = arrayOfNulls<String>(mServices!!.size)
            servicesId = IntArray(mServices!!.size)
            var i = 0
            for (service in mServices!!) {
                names[i] = service.name
                servicesId[i] = service.id
                i++
            }
            initServicesAdapter(names)
        }

        override fun onFailure(call: Call<ArrayList<Service>?>, t: Throwable) {
            println(t)
            Log.i(TAG, t.message!!)
            loadingDialog.dismiss()
        }
    }

    var roomsCallback: Callback<ArrayList<Placement>?> = object : Callback<ArrayList<Placement>?> {
        override fun onResponse(call: Call<ArrayList<Placement>?>, response: Response<ArrayList<Placement>?>) {
            if (!response.isSuccessful) {
                println(response.code())
                loadingDialog.dismiss()
                return
            }
            mPlacements = response.body()
            val ids = arrayOfNulls<Int>(mPlacements!!.size)
            var i = 0
            for (placement in mPlacements!!) {
                ids[i] = placement.id
                i++
            }
            initRoomsAdapter(ids)
        }

        override fun onFailure(call: Call<ArrayList<Placement>?>, t: Throwable) {
            println(t)
            Log.i(TAG, t.message!!)
            loadingDialog.dismiss()
        }
    }

    private fun initServicesAdapter(servicesData: Array<String?>) {
        val adapter: ArrayAdapter<String?> = ArrayAdapter<String?>(this,
                android.R.layout.simple_list_item_1, servicesData)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        mServicesSpinner.adapter = adapter
        loadingDialog.dismiss()
    }

    private fun initRoomsAdapter(roomsData: Array<Int?>) {
        val adapter: ArrayAdapter<Int?> = ArrayAdapter<Int?>(this,
                android.R.layout.simple_list_item_1, roomsData)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        mRoomsSpinner.adapter = adapter
        loadingDialog.dismiss()
    }

    @RequiresApi(Build.VERSION_CODES.N)
    private fun signContract() {
        loadingDialog.start()
        mContract!!.providerServiceId = servicesId[serviceId]
        mContract!!.placementId = roomId
        val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSX", resources.configuration.locales.get(0))
        mContract!!.date = dateFormat.format(Date(dateAndTime.timeInMillis))
        NetworkService.getInstance()
                .apiService
                .signContract(token, mContract)
                .enqueue(signContractCallback)
    }

    var signContractCallback: Callback<Contract?> = object : Callback<Contract?> {
        override fun onResponse(call: Call<Contract?>, response: Response<Contract?>) {
            if (!response.isSuccessful) {
                println(response.code())
                loadingDialog.dismiss()
                return
            }
            loadingDialog.dismiss()
            finish()
        }

        override fun onFailure(call: Call<Contract?>, t: Throwable) {
            println(t)
            Log.i(TAG, t.message!!)
            loadingDialog.dismiss()
        }
    }

    fun setDate(v: View?) {
        DatePickerDialog(
            this@SignContractActivity, d,
            dateAndTime[Calendar.YEAR],
            dateAndTime[Calendar.MONTH],
            dateAndTime[Calendar.DAY_OF_MONTH]
        )
            .show()
    }

    fun setTime(v: View?) {
        TimePickerDialog(
            this@SignContractActivity, t,
            dateAndTime[Calendar.HOUR_OF_DAY],
            dateAndTime[Calendar.MINUTE], true
        )
            .show()
    }

    var d =
        OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
            dateAndTime[Calendar.YEAR] = year
            dateAndTime[Calendar.MONTH] = monthOfYear
            dateAndTime[Calendar.DAY_OF_MONTH] = dayOfMonth
            setInitialDateTime()
        }

    var t =
        OnTimeSetListener { view, hourOfDay, minute ->
            dateAndTime[Calendar.HOUR_OF_DAY] = hourOfDay
            dateAndTime[Calendar.MINUTE] = minute
            setInitialDateTime()
        }

    private fun setInitialDateTime() {
        currentDateTime!!.text = DateUtils.formatDateTime(
            this,
            dateAndTime.timeInMillis,
            DateUtils.FORMAT_SHOW_DATE or DateUtils.FORMAT_SHOW_YEAR
                    or DateUtils.FORMAT_SHOW_TIME
        )
    }

    private fun navigateToMenuScreen() {
        val intent = Intent(this@SignContractActivity, MenuActivity::class.java)
        startActivity(intent)
    }

    companion object {
        private const val TAG = "SignContractActivity"
    }
}
